File: 0142-linked-list-cycle-ii/hashset/java/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        HashSet<ListNode> visited = new HashSet<>();
        ListNode cur = head;
        while (cur != null) {
            if (!visited.add(cur)) 
                return cur;     // Cycle entry found
            cur = cur.next;
        }
        return null;            // No cycle
    }
}

File: 0142-linked-list-cycle-ii/hashset/cpp/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode*> visited;
        ListNode* cur = head;
        while (cur) {
            if (visited.find(cur) != visited.end()) {
                return cur;     // Cycle entry found
            }
            visited.insert(cur);
            cur = cur->next;
        }
        return nullptr;        // No Cycle 
    }
};
